Flaky Test: `TestStoreGateway_InitialSyncWithWaitRingStability`

**Describe the bug**

The `TestStoreGateway_InitialSyncWithWaitRingStability` is flaky. [See this CI run](https://github.com/cortexproject/cortex/pull/4288/checks?check_run_id=2852832615):

```
 --- FAIL: TestStoreGateway_InitialSyncWithWaitRingStability (140.66s)
    --- FAIL: TestStoreGateway_InitialSyncWithWaitRingStability/shuffle_sharding_strategy,_20_gateways,_RF_=_3,_SS_=_3_(bucket_index_enabled_=_false) (32.45s)
        gateway_test.go:300: random generator seed: 1623962900927129822
        gateway_test.go:347: 
            	Error Trace:	gateway_test.go:347
            	Error:      	Received unexpected error:
            	            	invalid service state: Failed, expected: Running, failure: unable to start store-gateway dependencies: not healthy, 0 terminated, 1 failed: [register instance in the ring: failed to CAS store-gateway]
            	Test:       	TestStoreGateway_InitialSyncWithWaitRingStability/shuffle_sharding_strategy,_20_gateways,_RF_=_3,_SS_=_3_(bucket_index_enabled_=_false)
FAIL
FAIL	github.com/cortexproject/cortex/pkg/storegateway	219.005s
```

--- a/pkg/storegateway/gateway_test.go
+++ b/pkg/storegateway/gateway_test.go
@@ -300,7 +300,10 @@ func TestStoreGateway_InitialSyncWithWaitRingStability(t *testing.T) {
 				t.Log("random generator seed:", seed)
 
 				ctx := context.Background()
-				ringStore := consul.NewInMemoryClient(ring.GetCodec())
+				ringStore := consul.NewInMemoryClientWithConfig(ring.GetCodec(), consul.Config{
+					MaxCasRetries: 20,
+					CasRetryDelay: 500 * time.Millisecond,
+				})
 
 				// Create the configured number of gateways.
 				var gateways []*StoreGateway

